File: tsp.cpp
#include <vector>
#include <array>
#include <map>
#include <iostream>
#include <fstream>
#include <iterator>
#include <string>
#include <algorithm>
#include <cmath>

using namespace std;

/*

dynamic programming algorithm
construct 2D table : set of nodes in path, destination
base case : set of two elements, value equals the dist between them 
recurrence : shortest tour of elements in set excluding the destination plus distance from destination of shorter tour to final destination
loop over set sizes from 2 through 25
loop over all sets of the current size
loop over all destinations in the current set (the source node may not serve as the destination)
after table is filled out, loop over all problems of size 25, and add to each the distance from its destination node back to the source node.

need to generate combinations
use a map for the set/destination->distance table. (what kind of sequence objects can an array use as a key)


*/

/*
combination generating recursive algorithm
int k = 5; // k will range from 2 through 25
vector<array<int,k>> get_combos(range,choose_x){
	// only need k-1 items, then tack on 1.
	// need a forloop over 1..k-1
	vector <array<int,k>> combos;
	int start = 0;
	for (int el_id = 1; el_id < k; el_id++){
		// loop over all choices for first el
	}
}
<array<int,k> get_combo(range,choose_x,start_index){ // range = {2...25}, 
	
	for first_index 
	// choose element from range[start_index:range.size()-1]
	return {element, get_combo(range,choose_x-1,start_index = range.find(element)+1)} 			
}
*/

int main(){
	// put nodes in a vector
	ifstream infile("/home/damian/Dropbox/Algorithms/tsp_data.txt");
	vector<array<double,2>> node_coordinates;
	if (infile.good()){
		string skip_first_line;
		getline(infile,skip_first_line);
		int break_index = 10;
		for (int i = 1; i < 26; i++){
			string next_node;
			getline(infile,next_node);
			int x;
			int y;
			x = stod(next_node.substr(0,break_index));
			y = stod(next_node.substr(break_index+1));
			array<double,2> x_y = {x,y};
			node_coordinates.push_back(x_y);
		}
	}
	map<array<int,2>,double> edge_lengths;
	// put edges in a map
	for (int A = 0; A < 25; A++){
		for (int B = A+1; B < min(A+11,25); B++){
			array<double,2> first = node_coordinates[A];
			array<double,2> second = node_coordinates[B];
			array<int,2> f_s_pair = {A,B};
			array<int,2> s_f_pair = {B,A};
			double first_x = first[0];
			double first_y = first[1];
			double second_x = second[0];
			double second_y = second[1];
			double dist = sqrt(pow(first_x-second_x,2.0) + pow(first_y-second_y,2.0))
			edge_lengths[f_s_pair] = dist;
			edge_lengths[s_f_pair] = dist;
		}
	}
	// loop over set sizes
	for (int set_size = 2; set_size < 26; set_size++){
		// loop over sets of that size
			
	}
	

	// loop over destinations
}

File: tsp.py
#!/usr/env/bin python

'''
import numpy, itertools

tsp_data = numpy.genfromtxt('/home/damian/Dropbox/Algorithms/tsp_data.txt',skip_header=1)

# 

node_distances = numpy.zeros([25,25])

for i in range(25):
	for j in range(i,25):
		if i == j:
			node_distances[i][j] = 0
		else:
			node_distances[i][j] = numpy.sqrt((tsp_data[i][0] - tsp_data[j][0])**2 + (tsp_data[i][1] - tsp_data[j][1])**2)	
			node_distances[j][i] = node_distances[i][j]



global_min_cost = float("inf")

# do we have to cycle through all the start vertices? yeah i guess so...

for i in range(25):
	print "start node",i
	Answers = {}
	max_set = None
	for path_length in range(25):
		print "path length",path_length
		candidates = range(25)
		candidates.remove(i)
		sets = itertools.combinations(candidates,path_length)
		# for each set of nodes of this size, loop over all possible destinations
		counter = 0
		for Set in sets:
			counter += 1
			print "count: ",counter
			Set = list(Set)
			Set.append(i)
			if len(Set) == 25:
				max_set = Set
			for j in Set:
				if j == i and path_length == 0:
					Answers[(tuple(Set),j)] = 0
				elif j == i and path_length != 0:
					Answers[(tuple(Set),j)] = float("inf")
				else: 
					possible = [Answers[(tuple([x for x in Set if x!=j]),k)] + node_distances[j][k] for k in Set if k!=j]
					Answers[(tuple(Set),j)] = min(possible)		
	# loop over max_size set's possible destination nodes, tack on hop back to i. take the min and compare to global_min_cost
	distances = [Answers[(tuple(max_set),j)] + node_distances[i][j] for j in range(25)]					
	# round down to nearest integer
	cheapest = int(min(distances)//1)
	if cheapest < global_min_cost:
		global_min_cost = cheapest
'''



# for every start and destination pair ij, find the shortest path that uses each node in S exactly once. 

# what should it look like

# define the special cases, when the Set includes only the start vertex and the destination vertex == the start vertex.

# how do i build the set S? Includes source and destination

# set size C from 0 to 24, get combination of size C, tack on source vertex to each combination, loop over possible destinations

# recurrence : take destination out of set, randomly choose a new destination from set. Return dist_from_new_to_old_dest + shortest path to new.
# we want the minimum over all choices of new destination. 

# base cases
# if source == new_dest and S has only source inside it then dist = 0
# if source == new_dest and S has other stuff inside it then dist is INF  

# after done, need to loop over all source-dest pairs of size S = 25, tack on cost of final hop back to source, find the min of all those



# strategy for reducing problem-space

# 1.) Sets considered : all node indices must be distance <= 5 from each other
# 2.) subproblems from which to calculate problem : all node indices must be distance <= 5 from each other (AFTER new_dest k is removed) 


import numpy, itertools

def satisfies_constraint(the_set):
	'''
	Ensures that all consecutive indices are within 5 of one another

	the_set : list

	Returns : bool	
	'''
	for i in range(len(the_set)-1):
		if abs(the_set[i] - the_set[i+1]) > 5:
			return False
	else:
		return True
			

import copy

def satisfies_constraint_without(the_set,*args):
	'''
	Checks whether the set satisfies the constraint once k is removed

	args : int

	the_set : list

	Returns : bool
	'''
	safe = copy.copy(the_set) 
	for arg in args:
		safe.remove(arg)
	return satisfies_constraint(safe)	



tsp_data = numpy.genfromtxt('/home/damian/Dropbox/Algorithms/tsp_data.txt',skip_header=1)

# 

node_distances = numpy.zeros([25,25])

for i in range(25):
	for j in range(i,25):
		if i == j:
			node_distances[i][j] = 0
		else:
			node_distances[i][j] = numpy.sqrt((tsp_data[i][0] - tsp_data[j][0])**2 + (tsp_data[i][1] - tsp_data[j][1])**2)	
			node_distances[j][i] = node_distances[i][j]



global_min_cost = float("inf")

# do we have to cycle through all the start vertices? yeah i guess so...

for i in range(1):
	print "start node",i
	Answers = {}
	max_set = None
	for path_length in range(25):
		print "path length",path_length
		candidates = range(25)
		candidates.remove(i)
		sets = itertools.combinations(candidates,path_length)
		# for each set of nodes of this size, loop over all possible destinations
		counter = 0
		for Set in sets:
			counter += 1
			print 'count',counter
			Set = list(Set)
			Set.append(i)
			if len(Set) == 25:
				max_set = Set
			safe_copy = copy.copy(Set)
			safe_copy = sorted(safe_copy)
			if not satisfies_constraint(safe_copy):
				continue
			for j in Set:
				if not satisfies_constraint_without(safe_copy,j):
					continue
				if j == i and path_length == 0:
					Answers[(tuple(Set),j)] = 0
				elif j == i and path_length != 0:
					Answers[(tuple(Set),j)] = float("inf")
				else: 
					possible = [Answers[(tuple([x for x in Set if x!=j]),k)] + node_distances[j][k] for k in Set if k!=j and satisfies_constraint_without(safe_copy,j,k)]
					Answers[(tuple(Set),j)] = min(possible)		
	# loop over max_size set's possible destination nodes, tack on hop back to i. take the min and compare to global_min_cost
	distances = [Answers.get((tuple(max_set),j),float('inf')) + node_distances[i][j] for j in range(25)]					
	# round down to nearest integer
	cheapest = int(min(distances)//1)
	if cheapest < global_min_cost:
		global_min_cost = cheapest



'''
TSP review

read in data
store nodes in a vector of 2-arrays. 

loop over nodes, where indices <=10 of each other
store distances between pairs in a map. store both permutations of vertices ex. {(x,y) -> 5, (y,x) -> 5}




loop over set size (btwn 2 and 25)
	loop over sets of that size (each set must contain 1, and consecutive elements' indices must be within 10 of each other)
		loop over nodes in the set (each member except 1 gets to try out being the destination node)
			base case : set size is 2. Path length is distance between nodes. 
			otherwise use the recurrence: dist from j to k plus min path to j using nodes in S except k.


'''
		























